deepreg/model/loss/deform.py

```diff
@@ -1,49 +1,55 @@
-        return (fv[:, 1:-1, 2:, 1:-1] - fv[:, 1:-1, :-2, 1:-1]) / 2
-
-    def gradient_dz(fv):
-        return (fv[:, 1:-1, 1:-1, 2:] - fv[:, 1:-1, 1:-1, :-2]) / 2
-
-    def gradient_txyz(Txyz, fn):
-        return tf.stack([fn(Txyz[..., i]) for i in [0, 1, 2]], axis=4)
-
-    def compute_gradient_norm(displacement, l1=False):
-        dTdx = gradient_txyz(displacement, gradient_dx)
-        dTdy = gradient_txyz(displacement, gradient_dy)
-        dTdz = gradient_txyz(displacement, gradient_dz)
-        if l1:
-            norms = tf.abs(dTdx) + tf.abs(dTdy) + tf.abs(dTdz)
-        else:
-            norms = dTdx ** 2 + dTdy ** 2 + dTdz ** 2
-        return tf.reduce_mean(norms, [1, 2, 3, 4])
-
-    def compute_bending_energy(displacement):
-        dTdx = gradient_txyz(displacement, gradient_dx)
-        dTdy = gradient_txyz(displacement, gradient_dy)
-        dTdz = gradient_txyz(displacement, gradient_dz)
-        dTdxx = gradient_txyz(dTdx, gradient_dx)
-        dTdyy = gradient_txyz(dTdy, gradient_dy)
-        dTdzz = gradient_txyz(dTdz, gradient_dz)
-        dTdxy = gradient_txyz(dTdx, gradient_dy)
-        dTdyz = gradient_txyz(dTdy, gradient_dz)
-        dTdxz = gradient_txyz(dTdx, gradient_dz)
-        return tf.reduce_mean(
-            dTdxx ** 2
-            + dTdyy ** 2
-            + dTdzz ** 2
-            + 2 * dTdxy ** 2
-            + 2 * dTdxz ** 2
-            + 2 * dTdyz ** 2,
-            [1, 2, 3, 4],
-        )
+def gradient_dx(fv):
+    return (fv[:, 2:, 1:-1, 1:-1] - fv[:, :-2, 1:-1, 1:-1]) / 2
+
+
+def gradient_dy(fv):
+    return (fv[:, 1:-1, 2:, 1:-1] - fv[:, 1:-1, :-2, 1:-1]) / 2
+
+
+def gradient_dz(fv):
+    return (fv[:, 1:-1, 1:-1, 2:] - fv[:, 1:-1, 1:-1, :-2]) / 2
+
 
+def gradient_txyz(Txyz, fn):
+    return tf.stack([fn(Txyz[..., i]) for i in [0, 1, 2]], axis=4)
+
+
+def compute_gradient_norm(displacement, l1=False):
+    dTdx = gradient_txyz(displacement, gradient_dx)
+    dTdy = gradient_txyz(displacement, gradient_dy)
+    dTdz = gradient_txyz(displacement, gradient_dz)
+    if l1:
+        norms = tf.abs(dTdx) + tf.abs(dTdy) + tf.abs(dTdz)
+    else:
+        norms = dTdx ** 2 + dTdy ** 2 + dTdz ** 2
+    return tf.reduce_mean(norms, [1, 2, 3, 4])
+
+
+def compute_bending_energy(displacement):
+    dTdx = gradient_txyz(displacement, gradient_dx)
+    dTdy = gradient_txyz(displacement, gradient_dy)
+    dTdz = gradient_txyz(displacement, gradient_dz)
+    dTdxx = gradient_txyz(dTdx, gradient_dx)
+    dTdyy = gradient_txyz(dTdy, gradient_dy)
+    dTdzz = gradient_txyz(dTdz, gradient_dz)
+    dTdxy = gradient_txyz(dTdx, gradient_dy)
+    dTdyz = gradient_txyz(dTdy, gradient_dz)
+    dTdxz = gradient_txyz(dTdx, gradient_dz)
+    return tf.reduce_mean(
+        dTdxx ** 2
+        + dTdyy ** 2
+        + dTdzz ** 2
+        + 2 * dTdxy ** 2
+        + 2 * dTdxz ** 2
+        + 2 * dTdyz ** 2,
+        [1, 2, 3, 4],
+    )
+
+
+def local_displacement_energy(ddf, energy_type, **kwargs):
```

Comment:
Line 60 not tested: (raise Value Error). Can you add a pytest assertion test here for completeness?

Reply:
Sorry for the delay, I've added the doc strings and assertion test.